### ISSUE TITLE
Unsupported codec on server should fail with Unimplemented instead of attempting to use proto codec

### ISSUE THREAD

**Issue by dfawley** [grpc/grpc-go#1824]

**Comment by easwars** [comment]
@dfawley : What if we make this change, but have an env var that people can set to get the old behavior? Or, we can consider the other approach, where we throw a log saying "oh, we didnt find the codec, we are going to be using proto for now, but this change".

**Comment by dfawley** [comment]
Yes, I think we could start by logging a warning and then have an environment variable to revert to the old behavior for a few releases.

**Comment by easwars** [comment]
Added a warning here: https://github.com/grpc/grpc-go/pull/6658

**Comment by easwars** [comment]
I sent a PR to add a test for this scenario which seems to be have been missed when I cleaned up the codec tests.

In a follow-up PR, I will introduce a new env var and will gate the new behavior behind that env var.

@dfawley Bikeshedding for the env var:
- GRPC_GO_SERVER_STRICT_CODEC_CHECKING
- GRPC_GO_SERVER_REQUIRE_MATCHING_CODEC
- GRPC_GO_SERVER_REJECT_UNKNOWN_CODEC
- GRPC_GO_SERVER_UNIMPLEMENTED_FOR_UNKNOWN_CODEC
- GRPC_GO_SERVER_DISABLE_PROTO_CODEC_FALLBACK

**Comment by dfawley** [comment]
`GRPC_GO_EXPERIMENTAL_PROTO_CODEC_FALLBACK` ?

I think "experimental" should be in the name, or something else to suggest it's supposed to be temporary.  And/or the key/value should include 1824 (this issue number) somehow.
